test/lib/web_device_interface.js

```diff
@@ -79,16 +79,30 @@ export default class WebDeviceInterface extends DeviceInterface {
     }
   }
 
-  /** @keychian */
-  async getKeychainValue() {
-    return this.keychainValue;
+  /** @keychain */
+  async getNamespacedKeychainValue() {
+    return this.keychainValue[this.keychainStorageKey];
```

Comment:
So actually I'm thinking that `keychainStorageKey` is no longer required (in general). Instead, this function would just return `return this.keychainValue[this.namespace];` Because the implementation of keychainStorageKey is:

```javascript
  protected get keychainStorageKey() {
    const keychainStorageKey = 'keychain';
    if (this.namespace) {
      return `${this.namespace}-${keychainStorageKey}`;
    } else {
      return keychainStorageKey;
    }
  }
```

But, in what cases would namespace be undefined? It should never be undefined.

I think essentially we got confused with a lot of different concepts. This should be the final structure:

- Each client stores one main keychain value with the storage key `keychain`. For web, this value would be stored in localStorage. For mobile, this value would be stored in the device keychain.
- The `keychain` value is a map: [namespace : value].

Reply:
Also, anytime we call `setNamespacedKeychainValue`, we must also persist the root structure to the device keychain. Because `this.keychainValue` is just an in-memory structure.

Reply:
> But, in what cases would namespace be undefined? It should never be undefined.

The namespace is initialized as an empty string [here](https://github.com/standardnotes/web/blob/79d48b7e077394dcf577438bcc54720ceddb8a87/app/assets/javascripts/ui_models/application.ts#L59).

The below block of code will return 'keychain':

```javascript
protected get keychainStorageKey() {
  const keychainStorageKey = 'keychain';
  if (this.namespace) {
    return `${this.namespace}-${keychainStorageKey}`;
  } else {
    return keychainStorageKey;
  }
}
```

Reply:
But I get the general idea, it looks much better. Maybe we can validate the `namespace` to be required and not be an empty string? Also, we need to update the `web` and `mobile` apps to have a valid namespace.

Reply:
The empty string was essentially a placeholder. We need to dynamically determine an automatic namespace to use, based on how many existing namespaces there are. So, web should pass undefined as namespace, and at that point, SNJS should automatically assign a dynamic namespace that is persisted.

Let me know what you think are good ways to solve this. One idea:

- We could have a storage key `namespaces` that is an array of namespaces: [app-1, app-2, app-3]. Then when creating a new app, we retrieve the last namespace in that array, and increment it to get a new namespace: app-4.

You could build this functionality into this existing PR.

Reply:
We need to also consider the UX of this. If a user has multiple accounts signed into a client, and each account uses its own keychain/storage namespace, how will a user know which account they are switching into if the account data (email address, etc) is encrypted? One solution would be that each namespace stores a user-defined "label" that allows them to distinguish one namespace from the other. This value would not be encrypted and would be stored in root storage in the `namespaces` struct.

Reply:
We can also generate an automatic namespace by taking the length of `namespaces` and increment it. 

But when will SNJS automatically assign a dynamic namespace? When a new account is signed in?

Reply:
If namespaces.length == 0, create new namespace;
If namespaces.length == 1, clients should automatically sign in to that namespace;
If namespaces.length > 1, clients should display UI for selecting which namespace to sign into (this is post-3.5, maybe like 3.6 or 3.7).

So in general it would seem like each namespace value should be an object, so that we can have properties like `label`, `isDefault`, etc.

Reply:
I see. We will put this feature on hold for now, we'll discuss this further once we get back on this.